File: Sorting/Merge Sort/C++/MergeSort.cpp
/**
 * @file MergeSort.cpp
 *
 * @brief Implementation of Merge Sort in C++
 *
 * @author ashishlamsal
 *
 */

#include <iostream>

void merge(int left[], int right[], int result[], int l1, int l2) {
    int i = 0, j = 0, k = 0;

    while (i < l1 && j < l2) {
        if (left[i] <= right[j])
            result[k++] = left[i++];
        else
            result[k++] = right[j++];
    }

    while (i < l1)
        result[k++] = left[i++];

    while (j < l2)
        result[k++] = right[j++];
}

void merge_sort(int arr[], int n) {
    if (n < 2) return;

    // divide
    int middle = n / 2;
    int* left = new int[middle];
    for (int i = 0; i < middle; i++)
        left[i] = arr[i];

    int* right = new int[n - middle];
    for (int i = middle; i < n; i++)
        right[i - middle] = arr[i];

    // sort
    merge_sort(left, middle);
    merge_sort(right, n - middle);

    // conquer
    merge(left, right, arr, middle, n - middle);

    delete[] left;
    delete[] right;
}

// Sample Input:  [ 8 6 10 5 9 3 2 6 4 ]
// Sample Output: [ 2 3 4 5 6 6 8 9 10 ]

// TIME COMPLEXITY: O(nlog(n))
// SPACE COMPLEXITY: O(n)

File: Sorting/Merge Sort/Python/MergeSort.py
def merge(array: [int], beg: int, m: int, end: int):
    temp = [0]* (end - beg + 1)
    i,j,k = beg, m + 1, 0
    
    while i <= m and j <= end:
        if array[i] <= array[j]:
            temp[k] = array[i]
            k += 1
            i += 1
        else:
            temp[k] = array[j]
            k += 1
            j += 1     

    while i <= m:
        temp[k] = array[i]
        i += 1
        k += 1
 
    while j <= end:
        temp[k] = array[j]
        j += 1
        k += 1      

    for i in range(beg,end+1):
        array[i] = temp[i - beg]      

def Merge_sort(array: [int], beg: int, end: int):
    if beg < end:
        m = (beg + end)/2
        Merge_sort(array,beg,m)
        Merge_sort(array,m+1,end)
        merge(array,beg,m,end)

#Sample Input: 5 14 35 20 7 2
#Sample Output: 2 5 7 14 20 35
#Time Complexity: θ(nLogn) in all cases(worst, average and best)
#Space Complexity: O(n)  
